File: rydberg_sim_py/stark_zeeman.py
# -*- coding: utf-8 -*-
"""
Stark Map and Zeeman Splitting calculations for Hydrogen-like atomic Rydberg manifolds.
Part of the Research-Grade Rydberg AMO Simulation package.
Calculates Stark and Zeeman shifts and diagonalizes the coupled Hamiltonian.
"""

import numpy as np
from .atomic import E_CHARGE, BOHR_RADIUS, H_PLANCK, BOHR_MAGNETON, AtomModel

class StarkZeemanEngine:
    def __init__(self, atom_model: AtomModel):
        self.atom = atom_model

    def build_hamiltonian_mhz(self, n: int, m: int, e_field_v_m: float, b_field_gauss: float) -> tuple:
        """
        Builds the Rydberg Stark and Zeeman Hamiltonian in the |n, l, m> basis.
        Couples delta_l = +-1 states using dynamic radial dipole elements.
        
        Args:
            n: principal quantum number
            m: projection quantum number
            e_field_v_m: Electric field along z axis in V/m
            b_field_gauss: Magnetic field along z axis in Gauss
            
        Returns:
            H_mhz: Hamiltonian matrix (2D numpy array) in MHz
            basis_l: list of l-values matching the matrix row index
        """
        abs_m = abs(m)
        basis_l = list(range(abs_m, n))
        dim = len(basis_l)
        H = np.zeros((dim, dim), dtype=float)

        # 1 Gauss = 1e-4 Tesla
        b_tesla = b_field_gauss * 1e-4
        # Zeeman shift delta_E = m * mu_B * B / h (using g_L = 1 for singlet Rydberg states)
        zeeman_mhz = (BOHR_MAGNETON * b_tesla * m) / (H_PLANCK * 1e6)

        # Diagonal Elements: Atomic Energies (from defects) + Constant Zeeman shift
        # Note: Zeeman shift is identical for all states inside this m-conserving manifold!
        for i, l in enumerate(basis_l):
            l_char = self._get_l_char(l)
            energy_hz = self.atom.get_energy(n, l_char) / H_PLANCK
            H[i, i] = (energy_hz / 1e6) + zeeman_mhz

        # Off-Diagonal Elements: Stark Dipole Couplings
        # H_stark = -e * E * z, where z = r * cos(theta)
        for i in range(dim - 1):
            l = basis_l[i]
            next_l = basis_l[i + 1]
            
            # Selection rules: l -> l+1
            if next_l == l + 1:
                # 1. Angular matrix element: <l, m| cos(theta) | l+1, m>
                angular_term = np.sqrt(
                    (next_l**2 - m**2) / ((2 * l + 1) * (2 * next_l + 1))
                )
                
                # 2. Radial matrix element coupling: <n, l+1| r | n, l>
                # Hydrogenic exact radial matrix factor within same n manifold:
                # <n, l| r | n, l-1> = 1.5 * n * sqrt(n^2 - l^2) * a_0
                radial_term = 1.5 * n * np.sqrt(n**2 - next_l**2) * BOHR_RADIUS
                
                # Coupling coefficient in Joules
                coupling_j = -E_CHARGE * e_field_v_m * radial_term * angular_term
                coupling_mhz = coupling_j / (H_PLANCK * 1e6)
                
                H[i, i + 1] = coupling_mhz
                H[i + 1, i] = coupling_mhz # Hermiticity

        return H, basis_l

    def diagonalize(self, n: int, m: int, e_field_v_m: float, b_field_gauss: float) -> tuple:
        """
        Diagonalizes the Stark-Zeeman Hamiltonian.
        Returns sorted eigenvalues and eigenvectors.
        """
        H, basis_l = self.build_hamiltonian_mhz(n, m, e_field_v_m, b_field_gauss)
        eigenvalues, eigenvectors = np.linalg.eigh(H)
        return eigenvalues, eigenvectors, basis_l

    def generate_stark_map(self, n: int, m: int, e_fields: np.ndarray, b_field_gauss: float) -> np.ndarray:
        """
        Sweeps electric field values and returns 2D array of eigenvalues for Stark plotting.
        Rows: e_field sweeps, Columns: Eigenenergy values in MHz
        """
        num_fields = len(e_fields)
        num_states = n - abs(m)
        eigen_grid = np.zeros((num_fields, num_states))

        for idx, e_val in enumerate(e_fields):
            eig_vals, _, _ = self.diagonalize(n, m, e_val, b_field_gauss)
            eigen_grid[idx, :] = eig_vals

        return eigen_grid

    def _get_l_char(self, l: int) -> str:
        letters = ['S', 'P', 'D', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N']
        if l < len(letters):
            return letters[l]
        return 'G' # fallback to hydrogenic

File: src/physics/starkZeeman.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { BOHR_RADIUS, E_CHARGE, H_PLANCK, BOHR_MAGNETON } from "./constants";
import { getRydbergEnergy, Species } from "./defectTheory";

// Diagonalizes any small symmetric real matrix in JavaScript/TypeScript using the Jacobi Algorithm.
// This is numerically stable, fast, self-contained, and preserves physical eigenvectors.
export function diagonalizeJacobi(
  matrix: number[][]
): { eigenvalues: number[]; eigenvectors: number[][] } {
  const n = matrix.length;
  // Initialize eigenvalues to diagonal
  const d = new Array(n).fill(0);
  const v = Array.from({ length: n }, () => new Array(n).fill(0));
  
  for (let i = 0; i < n; i++) {
    d[i] = matrix[i][i];
    v[i][i] = 1.0;
  }

  const maxIterations = 500;
  const tolerance = 1e-12;

  for (let iter = 0; iter < maxIterations; iter++) {
    let maxOffDiag = 0;
    let p = 0;
    let q = 0;

    // Find the largest off-diagonal element
    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        const val = Math.abs(matrix[i][j]);
        if (val > maxOffDiag) {
          maxOffDiag = val;
          p = i;
          q = j;
        }
      }
    }

    if (maxOffDiag < tolerance) {
      break; // Convergence reached
    }

    const apq = matrix[p][q];
    const app = d[p];
    const aqq = d[q];

    // Compute rotation angle (theta)
    const phi = 0.5 * (aqq - app) / apq;
    let t = 1.0 / (Math.abs(phi) + Math.sqrt(1.0 + phi * phi));
    if (phi < 0) t = -t;

    const c = 1.0 / Math.sqrt(1.0 + t * t);
    const s = t * c;
    const tau = s / (1.0 + c);

    // Update diagonal values
    d[p] = app - t * apq;
    d[q] = aqq + t * apq;

    matrix[p][q] = 0;

    // Update matrix elements elements coupling to rotational space
    for (let i = 0; i < n; i++) {
      if (i !== p && i !== q) {
        const aip = matrix[i][p];
        const aiq = matrix[i][q];
        matrix[i][p] = aip - s * (aiq + aip * tau);
        matrix[p][i] = matrix[i][p];
        matrix[i][q] = aiq + s * (aip - aiq * tau);
        matrix[q][i] = matrix[i][q];
      }
    }

    // Update eigenvectors
    for (let i = 0; i < n; i++) {
      const vip = v[i][p];
      const viq = v[i][q];
      v[i][p] = vip - s * (viq + vip * tau);
      v[i][q] = viq + s * (vip - viq * tau);
    }
  }

  // Sort by eigenvalue (ascending order)
  const indices = Array.from({ length: n }, (_, i) => i);
  indices.sort((a, b) => d[a] - d[b]);

  const sortedEigenvalues = indices.map((i) => d[i]);
  const sortedEigenvectors = Array.from({ length: n }, () => new Array(n).fill(0));
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      sortedEigenvectors[i][j] = v[i][indices[j]];
    }
  }

  return { eigenvalues: sortedEigenvalues, eigenvectors: sortedEigenvectors };
}

/**
 * Builds the Hamiltonian matrix in the basis |n, l, m> for a given local n and m.
 * Handles the Stark Effect via electric field Ey, and the Zeeman Effect via magnetic field bField.
 * Base energies calculated via Species defects.
 */
export function buildStarkZeemanHamiltonian(
  species: Species,
  n: number,
  m: number,
  electricFieldV_m: number, // V / m
  magneticFieldGauss: number // Gauss (1 Gauss = 1e-4 T)
): { hamiltonianMhz: number[][]; basisL: number[] } {
  const absM = Math.abs(m);
  // Basis: l ranges from |m| up to n-1
  const basisL: number[] = [];
  for (let l = absM; l <= n - 1; l++) {
    basisL.push(l);
  }

  const dim = basisL.length;
  const H = Array.from({ length: dim }, () => new Array(dim).fill(0));

  // MHz conversions:
  // E_Mhz = E_Hz / 1e6
  // Zeeman frequency factor: mu_B * B_tesla / h
  // 1 Gauss = 10^-4 Tesla
  const bTesla = magneticFieldGauss * 1e-4;
  const zeemanShiftMhz = (BOHR_MAGNETON * bTesla * m) / (H_PLANCK * 1e6); // g_L is exactly 1 for singlet orbital, and m_j projection is m

  for (let i = 0; i < dim; i++) {
    const lValue = basisL[i];
    // Diagonal element: Atomic Rydberg energy for the state (in MHz) + Zeeman shift
    const stateEnergyHz = getRydbergEnergy(species, n, lNumToLetter(lValue)).energyHz;
    const stateEnergyMhz = stateEnergyHz / 1e6;
    
    // Total diagonal energy
    H[i][i] = stateEnergyMhz + zeemanShiftMhz;

    // Off-diagonal Stark couplings: <l | -e * E * z | l+1>
    // Transition coupling is selection-rule constrained delta_l = 1
    if (i < dim - 1) {
      const nextLValue = basisL[i + 1];
      if (nextLValue === lValue + 1) {
        // Stark electric field perturbation amplitude:
        // H_Stark = -e * E_field * <n, l, m | z | n, l+1, m>
        // Angular matrix element: <l, m | cos(theta) | l+1, m> = sqrt( ((l+1)^2 - m^2) / ((2l+1)*(2l+3)) )
        const angularTerm = Math.sqrt(
          (Math.pow(nextLValue, 2) - Math.pow(m, 2)) /
          ((2 * lValue + 1) * (2 * nextLValue + 1))
        );

        // Radial matrix element inside same n manifold:
        // <n, l | r | n, l-1> = 1.5 * n * sqrt(n^2 - l^2) * a_0
        // Here, we couple lValue to nextLValue = lValue + 1 direction:
        // radial dipole is <n, l+1 | r | n, l> = 1.5 * n * sqrt(n^2 - (l+1)^2) * a_0
        const radialTermInstance = 1.5 * n * Math.sqrt(Math.pow(n, 2) - Math.pow(nextLValue, 2)) * BOHR_RADIUS;

        // Transition coupling in Joules
        const couplingJ = -E_CHARGE * electricFieldV_m * radialTermInstance * angularTerm;
        // Convert to MHz frequency
        const couplingMhz = couplingJ / (H_PLANCK * 1e6);

        H[i][i + 1] = couplingMhz;
        H[i + 1][i] = couplingMhz; // Hermiticity
      }
    }
  }

  return { hamiltonianMhz: H, basisL };
}

/**
 * Generate Stark Maps (Eigenvalues under varying electric field)
 * Return structure is suitable for responsive Recharts plots.
 */
export interface StarkMapPoint {
  eField: number;
  [stateIndex: string]: number; // Eigenenergy of state in MHz
}

export function generateStarkMap(
  species: Species,
  n: number,
  m: number,
  eFields: number[], // List of fields to sweep
  bFieldGauss: number
): StarkMapPoint[] {
  const points: StarkMapPoint[] = [];

  for (const eField of eFields) {
    const { hamiltonianMhz, basisL } = buildStarkZeemanHamiltonian(
      species,
      n,
      m,
      eField,
      bFieldGauss
    );
    
    // Diagonalize the matrix at this specific field point
    const { eigenvalues } = diagonalizeJacobi(hamiltonianMhz);

    const point: StarkMapPoint = { eField };
    eigenvalues.forEach((energy, idx) => {
      point[`val_${idx}`] = energy;
    });

    points.push(point);
  }

  return points;
}

function lNumToLetter(lNum: number): string {
  const letters = ["S", "P", "D", "F", "G", "H", "I", "J", "K", "L", "M"];
  return letters[lNum] || `l=${lNum}`;
}
